services/m3dbnode/server/server.go

```diff
@@ -0,0 +1,748 @@
+	"github.com/m3db/m3db/retention"
+	m3dbruntime "github.com/m3db/m3db/runtime"
+	"github.com/m3db/m3db/services/m3dbnode/config"
+	"github.com/m3db/m3db/storage"
+	"github.com/m3db/m3db/storage/block"
+	"github.com/m3db/m3db/storage/cluster"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/storage/repair"
+	"github.com/m3db/m3db/storage/series"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/ts"
+	m3dbxio "github.com/m3db/m3db/x/io"
+	"github.com/m3db/m3db/x/tchannel"
+	xconfig "github.com/m3db/m3x/config"
+	"github.com/m3db/m3x/instrument"
+	xlog "github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/pool"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	bootstrapConfigInitTimeout = 10 * time.Second
+	serverGracefulCloseTimeout = 10 * time.Second
+)
+
+// RunOptions provides options for running the server
+// with backwards compatibility if only solely adding fields.
+type RunOptions struct {
+	// ConfigFile is the YAML configuration file to use to run the server.
+	ConfigFile string
+
+	// BootstrapCh is a channel to listen on to be notified of bootstrap.
+	BootstrapCh chan<- struct{}
+
+	// InterruptCh is a programmatic interrupt channel to supply to
+	// interrupt and shutdown the server.
+	InterruptCh <-chan error
+}
+
+// Run runs the server programmatically given a filename for the
+// configuration file.
+func Run(runOpts RunOptions) {
+	var cfg config.Configuration
+	if err := xconfig.LoadFile(&cfg, runOpts.ConfigFile); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to load %s: %v", runOpts.ConfigFile, err)
+	}
+
+	logger, err := cfg.Logging.BuildLogger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create logger: %v", err)
+	}
+
+	debug.SetGCPercent(cfg.GCPercentage)
+
+	scope, _, err := cfg.Metrics.NewRootScope()
+	if err != nil {
+		logger.Fatalf("could not connect to metrics: %v", err)
+	}
+
+	opts := storage.NewOptions()
+
+	iopts := opts.InstrumentOptions().
+		SetLogger(logger).
+		SetMetricsScope(scope).
+		SetMetricsSamplingRate(cfg.Metrics.SampleRate())
+	opts = opts.SetInstrumentOptions(iopts)
+
+	buildReporter := instrument.NewBuildReporter(iopts)
+	if err := buildReporter.Start(); err != nil {
+		logger.Fatalf("unable to start build reporter: %v", err)
+	}
+	defer buildReporter.Close()
+
+	configSvcClientOpts := cfg.ConfigService.NewOptions().
+		SetInstrumentOptions(
+			instrument.NewOptions().
+				SetLogger(logger).
+				SetMetricsScope(scope))
+	configSvcClient, err := etcdclient.NewConfigServiceClient(configSvcClientOpts)
+	if err != nil {
+		logger.Fatalf("could not create m3cluster client: %v", err)
+	}
+
+	dynamicOpts := namespace.NewDynamicOptions().
+		SetInstrumentOptions(iopts).
+		SetConfigServiceClient(configSvcClient).
+		SetNamespaceRegistryKey(kvconfig.NamespacesKey)
+	nsInit := namespace.NewDynamicInitializer(dynamicOpts)
+
+	opts = opts.
+		SetTickInterval(cfg.TickInterval).
+		SetNamespaceInitializer(nsInit)
+
+	newFileMode, err := cfg.Filesystem.ParseNewFileMode()
```

Comment:
super nit: could you move the cheaper validations above the more expensive ones. e.g. `newFileMode`/`hostid`/etc before connecting to kv. In the event of a bad config, it'd be good to fail earlier.

Reply:
Sure thing, done.